### PR TITLE
Test: Add stakingPool processExpiration unit tests

### PR DESCRIPTION
**Pull request by patitonar**
## Context

Closes https://github.com/NexusMutual/smart-contracts/issues/524

## Changes proposed in this pull request

Adds new unit tests for `processExpiration`. Followed the list of test cases from the issue

## Test plan

New tests were added in `test/unit/StakingPool/processExpirations.js`

## Checklist

- [x] Rebased the base branch
- [x] Attached corresponding Github issue
- [x] Prefixed the name with the type of change (i.e. feat, chore, test)
- [x] Performed a self-review of my own code
- [x] Followed the style guidelines of this project
- [ ] Made corresponding changes to the documentation
- [x] Didn't generate new warnings
- [x] Didn't generate failures on existing tests
- [x] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

**Review by shark0der — CHANGES_REQUESTED** [review]
The PR missed a bug caught by @kyledewy in his pricing tests. The bucket expirations has a tranche id misalignment that led to an underflow when attempting to expire the allocations from the wrong tranche.

The test that caught it is here: https://github.com/NexusMutual/smart-contracts/blob/3a818110bec135b874da8d9c5e9123cb85a299e5/test/unit/StakingPool/calculatePrice.js#L161

The bugfix is on #546. 

Please include some tests for that edge case.

File: test/unit/StakingPool/processExpirations.js [hunk 1]
```diff
@@ -108,4 +117,269 @@ describe('processExpirations', function () {
...
+    const accNxmPerRewardsShareAfter = await stakingPool.accNxmPerRewardsShare();
+    const rewardPerSecondAfter = await stakingPool.rewardPerSecond();
+    const lastAccNxmUpdateAfter = await stakingPool.lastAccNxmUpdate();
+    const expiredBucketRewards = await stakingPool.rewardBuckets(firstActiveBucketId);
+    const rewardsSharesSupply = await stakingPool.rewardsSharesSupply();
+
+    const bucketStartTime = firstActiveBucketId.mul(BUCKET_DURATION);
+    const elapsed = bucketStartTime.sub(lastAccNxmUpdateBefore);
+
+    expect(expiredBucketRewards).to.gt(0);
```
Comment (danoctavian) [inline test/unit/StakingPool/processExpirations.js:256]: can we have an exact value here for `expiredBucketRewards` ?

strengthen the assertion

File: test/unit/StakingPool/processExpirations.js [hunk 2]
```diff
@@ -108,4 +117,269 @@ describe('processExpirations', function () {
...
+    await stakingPool.processExpirations(true);
+
+    // Validate tranches are expired
+    for (let i = 0; i < tranches.length; i++) {
+      const tranche = tranches[i];
+      const expiredTranche = await stakingPool.expiredTranches(tranche);
+
+      const activeDepositsAtTranche = maxTranche - tranche + 1;
+
+      expect(expiredTranche.accNxmPerRewardShareAtExpiry).to.gt(0);
```
Comment (danoctavian) [inline test/unit/StakingPool/processExpirations.js:196]: let's see if we can turn this into an to.be.equal

to strengthen the assertion

Comment (patitonar) [reply test/unit/StakingPool/processExpirations.js:177]: I tried calculating the exact value but it was difficult to get it when all tranches expired at once without replicating the full contract logic from the `processExpirations()` method. That's why I left this assertion as it is and instead added a new test `'Expires tranches correctly storing expiredTranches struct` that expires one tranche at a time and validates the exact value for `expiredTranche.accNxmPerRewardShareAtExpiry` here https://github.com/NexusMutual/smart-contracts/pull/538/commits/328d5a5844af746ce1f0639ab9c679e855dc93f4

**Comment by patitonar** [comment]
> The PR missed a bug caught by @kyledewy in his pricing tests. The bucket expirations has a tranche id misalignment that led to an underflow when attempting to expire the allocations from the wrong tranche.
>The test that caught it is here:
>[smart-contracts/test/unit/StakingPool/calculatePrice.js](https://github.com/NexusMutual/smart-contracts/blob/3a818110bec135b874da8d9c5e9123cb85a299e5/test/unit/StakingPool/calculatePrice.js#L161)
>Line 161 in [3a81811](https://github.com/NexusMutual/smart-contracts/commit/3a818110bec135b874da8d9c5e9123cb85a299e5)
>The bugfix is on https://github.com/NexusMutual/smart-contracts/pull/546.
>Please include some tests for that edge case.

@shark0der It seems that the mentioned bug is related to the `requestAllocation()` method rather than the `processExpiration()` method. I agree that we should add tests for that edge case, but I think it would be better to include it as part of https://github.com/NexusMutual/smart-contracts/issues/492

File: test/unit/StakingPool/processExpirations.js (hunk 2 above)
Comment (danoctavian) [reply test/unit/StakingPool/processExpirations.js:177]: Thanks @patitonar 

Sounds ok


@shark0der let me know your thoughts here as well

**Review by danoctavian — APPROVED** [review]
LGTM

This looks fine for me now.

Benefits from a second look from @shark0der  as well.

File: test/unit/StakingPool/processExpirations.js [hunk 3]
```diff
@@ -88,4 +105,285 @@ describe('processExpirations', function () {
...
+    const { stakingPool } = this;
+    const {
+      members: [user],
+    } = this.accounts;
+
+    const { amount, tokenId, destination, initialPoolFee } = processExpirationsFixture;
+
+    const { firstActiveTrancheId, maxTranche } = await getTranches();
+
+    const tranches = Array(maxTranche - firstActiveTrancheId + 1)
```
Comment (shark0der) [inline test/unit/StakingPool/processExpirations.js:128]: Use the `MAX_ACTIVE_TRANCHES` constant instead

File: test/unit/StakingPool/processExpirations.js [hunk 4]
```diff
@@ -88,4 +105,285 @@ describe('processExpirations', function () {
...
+  it('anyone can call this method', async function () {
+    const { stakingPool } = this;
+    const {
+      nonMembers: [anyone],
+    } = this.accounts;
+
+    await expect(stakingPool.connect(anyone).processExpirations(true)).to.not.be.reverted;
+  });
+
+  it('Expires tranches updating active stake, stake shares and rewards shares supply', async function () {
```
Comment (shark0der) [inline test/unit/StakingPool/processExpirations.js:118]: Can you please lowercase the first letter in all `it('...')`s?

File: test/unit/StakingPool/processExpirations.js [hunk 5]
```diff
@@ -88,4 +105,285 @@ describe('processExpirations', function () {
...
+    let stakeSharesTotalSupply = parseEther('0');
+    for (let i = 0; i < tranches.length; i++) {
+      const tranche = tranches[i];
+      await stakingPool.connect(user).depositTo(amount, tranche, tokenId, destination);
+      const nftId = i + 1;
+      const deposit = await stakingPool.deposits(nftId, tranche);
+
+      const feesRewardShares = deposit.rewardsShares.mul(initialPoolFee).div(POOL_FEE_DENOMINATOR);
+
+      stakeSharesTotalSupply = stakeSharesTotalSupply.add(deposit.rewardsShares.add(feesRewardShares));
```
Comment (shark0der) [inline test/unit/StakingPool/processExpirations.js:141]: You're adding up reward shares but storing them in a "stake shares" variable.

File: test/unit/StakingPool/processExpirations.js [hunk 6]
```diff
@@ -88,4 +105,285 @@ describe('processExpirations', function () {
...
+    const { stakingPool } = this;
+    const {
+      members: [user],
+    } = this.accounts;
+
+    const { amount, tokenId, destination } = processExpirationsFixture;
+
+    const { firstActiveTrancheId, maxTranche } = await getTranches();
+
+    const tranchesAmount = maxTranche - firstActiveTrancheId + 1;
```
Comment (shark0der) [inline test/unit/StakingPool/processExpirations.js:208]: import the constant from helpers

**Review by shark0der — CHANGES_REQUESTED** [review]
Looks good overall! Minor changes requested.

File: test/unit/StakingPool/processExpirations.js (hunk 2 above)
Comment (shark0der) [reply test/unit/StakingPool/processExpirations.js:177]: Kinda torn on this one. How about implementing a test where we calculate the `accNxmPerRewardShare` when `processExpirations` expires `1 bucket + 1 tranche + 1 bucket` (in this order)? This way we can limit the number of calculations needed in the test itself but make sure that the calculation is done correctly.

**Comment by shark0der** [comment]
After requesting changes I've added [this comment](https://github.com/NexusMutual/smart-contracts/pull/538#discussion_r1060607163).

File: test/unit/StakingPool/processExpirations.js (hunk 2 above)
Comment (patitonar) [reply test/unit/StakingPool/processExpirations.js:178]: Great suggestion @shark0der ! I created a that new test `correctly calculates accNxmPerRewardShare` here [Add test for accNxmPerRewardShare calculation](https://github.com/NexusMutual/smart-contracts/pull/538/commits/6047d7b15751a5d71fba22daf9c3df2911229eb7)

**Review by shark0der — APPROVED** [review]
